src/wutil/dir_iter.rs

```diff
@@ -267,14 +266,16 @@ impl DirIter {
 
         // dent.d_name is c_char; pretend it's u8.
         assert!(std::mem::size_of::<libc::c_char>() == std::mem::size_of::<u8>());
-        let d_name_cchar = &dent.d_name;
-        let d_name = unsafe {
-            slice::from_raw_parts(d_name_cchar.as_ptr() as *const u8, d_name_cchar.len())
-        };
+
+        // TODO: This should use &raw from Rust 1.82 on
+        // https://github.com/rust-lang/libc/issues/2669
+        // https://github.com/fish-shell/fish-shell/issues/11221
+        let d_name_ptr = addr_of!((*dent).d_name);
+        let d_name = unsafe { std::ffi::CStr::from_ptr(d_name_ptr.cast()) }.to_bytes_with_nul();
```

Comment:
It's probably worth it to explain what's going on here - we can't trust the length (because dirent lies, a long-standing kernel C convention), but we can trust that it is NUL-terminated,
with a link to the libc issue.